sshmux_test.go

```diff
@@ -51,9 +76,86 @@ func sshAPIHandler(w http.ResponseWriter, r *http.Request) {
 func initHttp() {
 	http.HandleFunc("/ssh", sshAPIHandler)
 
-	if err := http.ListenAndServe("127.0.0.1:5000", nil); err != nil {
+	if err := http.ListenAndServe(apiServerAddr.String(), nil); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func initUpstreamProxyServer() {
+	listener, err := net.ListenTCP("tcp", sshmuxProxyAddr)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer listener.Close()
+
+	for {
+		conn, err := listener.Accept()
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		go func() {
+			// 1. Set up downstream connection with sshmux
+			sshmux, err := net.DialTCP("tcp", nil, sshmuxProxiedAddr)
+			if err != nil {
+				log.Fatal(err)
+			}
+			// 2. Send PROXY header to sshmux
+			header := proxyproto.HeaderProxyFromAddrs(2, conn.RemoteAddr(), nil)
+			_, err = header.WriteTo(sshmux)
+			if err != nil {
+				log.Fatal(err)
+			}
+			// 3. Forward TCP messages in both ways
+			go func() {
+				defer sshmux.Close()
+				io.Copy(sshmux, conn)
+			}()
+			go func() {
+				defer sshmux.Close()
```

Comment:
Closing `sshmux` twice?

Reply:
Well, in this way we’re shutting down the proxy connection in both direction (here `sshmux` is the upstream channel with sshmux held by the proxy, not the sshmux instance itself), i.e. if upstream `sshmux` closed the connection, the proxy will also  close the downstream connection with the client, vice versa.

Reply:
@iBug a48db81b685de926388573235180db6691610430 Is this what you mean…?